src/lang/internals/sclang-io.js

```diff
@@ -33,70 +36,83 @@ export const STATES = {
  */
 export class SclangIO extends EventEmitter {
   states: Object;
-  responseCollectors: Object;
-  capturing: Object;
   calls: Object;
   state: ?string;
-  output: Array<string>;
   result: Object;
+  partialMsg: string; // holds any message data we've received so far
+  partialStdout: string; // holds any leftover STDOUT data in-between data callbacks
+  captured: Array<string>; // holds the in-progress capture
+  captureGUID: string;
 
   constructor() {
     super();
     this.states = this.makeStates();
     this.reset();
   }
 
+  /*
+   * Handle a block of TCP data representing the response data from SCLang. We
+   * don't make any assumptions about the alignment of the callback data with
+   * the messages, and just look for a 0xff byte which indicates the end of a
+   * message (which is just a JSON string).
+   */
+  handleTCPData(data) {
+    var handledChars = 0;
+    var frameEnd = data.indexOf(FRAME_BYTE);
+    while(frameEnd !== -1) {
+      // we've got the end of a message - handle it
+      var msg = this.partialMsg + data.slice(handledChars, frameEnd);
+      this.partialMsg = '';
+      this.handleMsg(JSON.parse(msg));
+      handledChars = frameEnd+1; // also swallow the divider char
+      // look for another message
+      frameEnd = data.indexOf(FRAME_BYTE, handledChars);
+    }
+    // we've handled all complete messages, store any leftover data
+    this.partialMsg += data.slice(handledChars);
+  }
+
   reset() {
-    this.responseCollectors = {};
-    this.capturing = {};
+    this.partialMsg = '';
+    this.partialStdout = '';
+    this.captured = [];
+    this.captureGUID = '';
     this.calls = {};
     this.state = null;
-    // these are stored on the object
+    // this is stored on the object
     // and are sent with compile error/success event
-    this.output = [];
     this.result = {};
   }
 
   /**
-  * @param {string} input - parse the stdout of supercollider
+  * @param {string} input - parse the stdout of supercollider, line-by-line
   */
   parse(input: string) {
-    var echo = true, startState = this.state, last = 0;
-    this.states[this.state].forEach(stf => {
-      var match;
-      if (this.state === startState) {
-        while ((match = stf.re.exec(input)) !== null) {
-          last = match.index + match[0].length;
-          // do not post if any handler returns true
-          if (stf.fn(match, input) === true) {
-            echo = false;
-          }
-
-          // break if its not a /g regex with multiple results
-          if (!stf.re.global) {
-            break;
-          }
+    var inputLines = input.split('\n');
```

Comment:
I can see that parsing line by line is possibly better than multiline matching. I can also see a cleaner way to do it (than either my previous implementation or the one in this PR). But by changing this as much as you did, I'm now confused and can't quite tell what might break.

Yes, the tests run, but the reason I hadn't merged before is because I can't be sure now.